### PR TITLE
Refactor: make last_log_id an Option<LogId>

### LINKED ISSUE #49: Make `last_log_id` an `Option<LogId>`, or it can not express an uninitialized state
https://github.com/datafuselabs/openraft/blob/344d444c4ef84867d2ef09fb58032832b2148fcf/openraft/src/core/mod.rs#L140

### PR DESCRIPTION
**Pull request by YangKian**
Close #49

### CODE REVIEW

**Comment by YangKian** [comment]
I‘m done, PTAL~

File: openraft/src/core/append_entries.rs [hunk 1]
```diff
@@ -29,7 +29,8 @@ impl<D: AppData, R: AppDataResponse, N: RaftNetwork<D>, S: RaftStorage<D, R>> Ra
         &mut self,
         msg: AppendEntriesRequest<D>,
     ) -> Result<AppendEntriesResponse, AppendEntriesError> {
-        tracing::debug!(%self.last_log_id, %self.last_applied, msg=%msg.summary(), "handle_append_entries_request");
+        tracing::debug!(last_log_id=%self.last_log_id.map_or_else(|| "None".to_string(), |id| id.index.to_string()),
```
Comment (drmingdrmer) [inline openraft/src/core/append_entries.rs:32]: `?last_log_id` to print in debug mode.

File: openraft/src/core/append_entries.rs [hunk 2]
```diff
@@ -204,15 +205,17 @@ impl<D: AppData, R: AppDataResponse, N: RaftNetwork<D>, S: RaftStorage<D, R>> Ra
...
+                return Ok(());
+            }
         }
 
         tracing::debug!(
             "delete inconsistent log entries [{}, {}), last_log_id: {}, entries: {}",
             msg_entries[0].log_id,
             msg_entries[l - 1].log_id,
-            self.last_log_id,
+            self.last_log_id.map_or_else(|| "None".to_string(), |log_id| log_id.to_string()),
```
Comment (drmingdrmer) [inline openraft/src/core/append_entries.rs:218]: replace `last_log_id: {}` with `last_log_id: {:?}` to print an `Option<LogId>` directly.

File: openraft/src/metrics.rs [hunk 3]
```diff
@@ -56,11 +56,15 @@ pub struct RaftMetrics {
 
 impl MessageSummary for RaftMetrics {
     fn summary(&self) -> String {
+        let last_log_index = match self.last_log_index {
+            Some(idx) => idx.to_string(),
+            None => "None".to_string(),
+        };
         format!("Metrics{{id:{},{:?}, term:{}, last_log:{}, last_applied:{}, leader:{:?}, membership:{}, snapshot:{}, replication:{}",
```
Comment (drmingdrmer) [inline openraft/src/metrics.rs:63]: use `{:?}` to print Option

File: openraft/tests/stepdown.rs [hunk 4]
```diff
@@ -124,7 +125,10 @@ async fn stepdown() -> Result<()> {
     // As leader established it commits a Blank log.
     // If the election takes only one round, the expected term/index is 2/4.
     tracing::info!("term: {}", metrics.current_term);
-    tracing::info!("index: {}", metrics.last_log_index);
+    tracing::info!(
+        "index: {}",
+        metrics.last_log_index.map_or_else(|| "None".to_string(), |index| index.to_string())
```
Comment (drmingdrmer) [inline openraft/tests/stepdown.rs:130]: `{:?}`

File: openraft/tests/stepdown.rs [hunk 5]
```diff
@@ -92,9 +92,10 @@ async fn stepdown() -> Result<()> {
             metrics.current_term
         );
         assert_eq!(
-            metrics.last_log_index, 3,
+            metrics.last_log_index,
+            Some(3),
             "expected old leader to have last log index of 3, got {}",
-            metrics.last_log_index
+            metrics.last_log_index.map_or_else(|| "None".to_string(), |index| index.to_string())
```
Comment (drmingdrmer) [inline openraft/tests/stepdown.rs:98]: `{:?}`

File: openraft/tests/fixtures/mod.rs [hunk 6]
```diff
@@ -526,9 +526,11 @@ impl RaftRouter {
                 node.id, node.last_applied
             );
             assert_eq!(
-                node.last_log_index, 0,
+                node.last_log_index,
+                Some(0),
                 "node {} has last_log_index {}, expected 0",
-                node.id, node.last_log_index
+                node.id,
+                node.last_log_index.map_or_else(|| "None".to_string(), |index| index.to_string())
```
Comment (drmingdrmer) [inline openraft/tests/fixtures/mod.rs:533]: `{:?}`

File: openraft/tests/fixtures/mod.rs [hunk 7]
```diff
@@ -601,14 +604,20 @@ impl RaftRouter {
...
             );
             assert_eq!(
-                node.last_applied, expected_last_log,
+                Some(node.last_applied),
+                expected_last_log,
                 "node {} has last_applied {}, expected {}",
-                node.id, node.last_applied, expected_last_log
+                node.id,
+                node.last_applied,
+                expected_last_log.map_or_else(|| "None".to_string(), |log| log.to_string())
```
Comment (drmingdrmer) [inline openraft/tests/fixtures/mod.rs:612]: `{:?}`

File: openraft/tests/fixtures/mod.rs [hunk 8]
```diff
@@ -601,14 +604,20 @@ impl RaftRouter {
...
             );
             assert_eq!(
-                node.last_log_index, expected_last_log,
+                node.last_log_index,
+                expected_last_log,
                 "node {} has last_log_index {}, expected {}",
-                node.id, node.last_log_index, expected_last_log
+                node.id,
+                node.last_log_index.map_or_else(|| "None".to_string(), |index| index.to_string()),
+                expected_last_log.map_or_else(|| "None".to_string(), |index| index.to_string())
```
Comment (drmingdrmer) [inline openraft/tests/fixtures/mod.rs:620]: `{:?}`

**Review by drmingdrmer — COMMENTED** [review]
Nice shot!

The debug print for `Option` can be simplified with `{:?}`

**Review by lichuang — APPROVED** [review]

**Comment by drmingdrmer** [comment]
It will be merged automatically if there is one approved. Converted to draft until comments are resolved.

**Review by drmingdrmer — APPROVED** [review]
LGTM

**Comment by drmingdrmer** [comment]
@YangKian make it ready for review if it's been ready~:DDD 

Good job!

**Comment by drmingdrmer** [comment]
> so fast, I just want to confirm the test pass XD，thanks for all your help~

You do not need to pass the CI before marking it as ready.

**Pull request merged**